<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        // var a = { name: "Sam" };
        // var b = { name: "Tom" };
        // var o = {};
        // o[a] = 1;
        // o[b] = 2;
        // o['[object Object]']=3
        // console.log(o[a],o);


        // async function async1() {
        //     console.log("async1 start");
        //     await async2();
        //     console.log("async1 end");
        // }
        // async function async2() {
        //     console.log("async2");
        // }
        // console.log("script start");
        // setTimeout(() => {
        //     console.log("setTimeout");
        // }, 0);
        // async1();
        // new Promise((resolve) => {
        //     console.log("promise1");
        //     resolve();
        // }).then(() => {
        //     console.log("promise2");
        // });
        // console.log("script end");


        // const promise1 = Promise.resolve("First");
        // const promise2 = Promise.resolve("Second");
        // const promise3 = Promise.reject("Third");
        // const promise4 = Promise.resolve("Fourth");
        // const runPromises = async () => {
        //     const res1 = await Promise.all([promise1, promise2]);
        //     const res2 = await Promise.all([promise3, promise4]);
        //     console.log({res1,res2});

        //     return [res1, res2];
        // };
        // runPromises()
        //     .then((res) => console.log(res))
        //     .catch((err) => console.log(err));

        const p1 = new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve(1);
                throw new Error();
            }, 2000);
        });
        const p2 = p1
            .then((val) => {
                console.log(val);
                return val + 1;
            })
            .catch((err) => {
                console.log(err);
                return err;
            });
        Promise.all([p2, Promise.reject(3)])
            .then((val2) => {
                console.log(val2);
            })
            .catch((err2) => {
                console.log(err2);
            });
    </script>
</body>

</html>